File: Packets/src/main/java/me/ImSpooks/core/packets/init/channels/WrappedOutputStream.java
package me.ImSpooks.core.packets.init.channels;

import lombok.Getter;
import me.ImSpooks.core.helpers.Global;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.*;

/**
 * Created by Nick on 30 Sep 2019.
 * Copyright © ImSpooks
 */
public class WrappedOutputStream {

    @Getter private final List<Object> out;

    public WrappedOutputStream() {
        this.out = new ArrayList<>();
    }

    public void write(byte[] data) throws IOException {
        this.out.add(data);
    }

    public void write(int byt) throws IOException {
        this.out.add(byt);
    }

    public void writeBoolean(boolean b) throws IOException {
        this.out.add(b);
    }

    public void writeDouble(double d) throws IOException {
        this.out.add(d);
    }

    public void writeFloat(float f) throws IOException {
        this.out.add(f);
    }

    public void writeInt(int i) throws IOException {
        this.out.add(i);
    }

    public void writeLong(long l) throws IOException {
        this.out.add(l);
    }

    public void writeShort(short s) throws IOException {
        this.out.add(s);
    }

    public void writeString(String s) throws IOException {
        this.out.add(s);
    }

    public void writeFile(File file) throws IOException {
        byte[] fileBytes = Files.readAllBytes(file.toPath());
        Files.write(file.toPath(), fileBytes);
        this.write(fileBytes);
    }

    public void writeTypePrefixed(Object o) throws IOException {
        if (o == null) {
            this.write(-1);
        }
        else if (o instanceof String) {
            this.write(0);
            this.writeString((String) o);
        }
        else if (o instanceof Integer) {
            this.write(1);
            this.writeInt((Integer) o);
        }
        else if (o instanceof Long) {
            this.write(2);
            this.writeLong((Long) o);
        }
        else if (o instanceof Double) {
            this.write(3);
            this.writeDouble((Double) o);
        }
        else if (o instanceof Float) {
            this.write(4);
            this.writeFloat((Float) o);
        }
        else if (o instanceof Boolean) {
            this.write(5);
            this.writeBoolean((Boolean) o);
        }
        else if (o instanceof byte[]) {
            this.write(6);
            this.writeInt(((byte[]) o).length);
            this.out.add(o);
        }
        else if (o instanceof List) {
            List<?> list = (List<?>) o;
            this.write(7);
            this.writeInt(list.size());
            for (Object inList : list) {
                this.writeTypePrefixed(inList);
            }
        }
        else if (o instanceof Map) {
            HashMap<?, ?> map = (HashMap<?, ?>) o;
            if (o instanceof LinkedHashMap)
                this.write(8);
            else this.write(9);
            this.writeString(Global.GSON.toJson(map));
        }
        else {
            throw new UnsupportedOperationException(String.format("Cannot write data with class \'%s\'", o.getClass().getSimpleName()));
        }
    }
}

File: ClientBridge/src/main/php/client/packets/init/channels/WrappedOutputStream.php
<?php

namespace client\packets\init\channels;

use ArrayUtils;
use UnexpectedValueException;

/**
 * Created by Nick on 14 okt. 2019.
 * Copyright © ImSpooks
 */
class WrappedOutputStream {
    // TODO convert to 7.4 if released

    /** @var array */ private $out = [];

    public function write($bytes) {
        array_push($this->out, $bytes);
    }

    public function writeBoolean(bool $b) {
        array_push($this->out, $b);
    }

    public function writeDouble(float $d) {
        array_push($this->out, $d);
    }

    public function writeFloat(float $f) {
        array_push($this->out, $f);
    }

    public function writeInt(int $i) {
        array_push($this->out, $i);
    }

    public function writeLong(int $l) {
        array_push($this->out, $l);
    }

    public function writeShort(int $s) {
        array_push($this->out, $s);
    }

    public function writeString(String $s) {
        array_push($this->out, $s);
    }

    public function writeTypePrefixed($o) {
        if ($o == null) {
            $this->write(-1);
        }
        else if (is_string($o)) {
            $this->write(0);
            $this->writeString($o);
        }
        else if (is_int($o)) {
            $this->write(1);
            $this->writeInt($o);
        }
        else if (is_long($o)) {
            $this->write(2);
            $this->writeLong($o);
        }
        else if (is_double($o)) {
            $this->write(3);
            $this->writeDouble((Double) $o);
        }
        else if (is_float($o)) {
            $this->write(4);
            $this->writeFloat((Float) $o);
        }
        else if (is_bool($o)) {
            $this->write(5);
            $this->writeBoolean($o);
        }
        else if (is_array($o)) {
            if (!ArrayUtils::isMap($o)) { // list
                $this->write(7);
                $this->writeInt(sizeof($o));
                foreach($o as $inList)
                    $this->writeTypePrefixed($inList);
            }
            else { // map
                $this->write(8);
                $this->writeString(json_encode($o));
            }
        }
        else {
            throw new UnexpectedValueException(sprintf("Cannot write data with class \'%s\'", gettype($o)));
        }
    }

    public function getOut(): array {
        return $this->out;
    }
}